Expand Impact of Optimize sizing guidance with variable-filtering findings (#9227)
camunda/camunda-docs: docs/components/best-practices/architecture/sizing-your-environment.md

Optimize is an optional component that provides process analytics and reporting. When enabled, it has significant implications for sizing.

The figures below come from Camunda 8.9 load tests. Because 8.8 and 8.9 share the same exporter architecture, they apply to 8.8+ as well.

#### In short

- Enabling Optimize roughly **triples to quadruples Elasticsearch CPU and disk** at a realistic workload (around 3.4x CPU, 3.6x disk), largely independent of throughput.

> - It lowers achievable **throughput by 25-50%** on the same hardware.

Author: ```suggestion
- It lowers achievable **processing throughput by 25-50%** on the same hardware.
```

- The single most effective mitigation is to **keep variables out of Optimize**: this recovers around 60% of the storage and 65% of the CPU, plus most of the lost throughput, at the cost of variable-based analytics.
- Size Elasticsearch/OpenSearch accordingly (CPU, disk, **and shard budget**), or run Optimize on a **dedicated Elasticsearch/OpenSearch instance**.

For how Optimize fits into the export pipeline, see [Optimize data flow](./data-flow.md#optimize-data-flow). The full studies behind these numbers are [Impact of Optimize on Camunda](https://camunda.github.io/zeebe-chaos/2026/06/10/Impact-of-Optimize-on-Camunda) and [Reducing Optimize's Elasticsearch overhead](https://camunda.github.io/zeebe-chaos/2026/06/25/Impact-of-Optimize-Variable-Filtering).

#### Why Optimize matters for sizing

- Optimize is a second-tier consumer of the export pipeline: the Elasticsearch/OpenSearch exporter writes raw engine events, Optimize's importer reads them and writes its own analytics indices back to Elasticsearch/OpenSearch, so data is written to secondary storage twice. See [Optimize data flow](./data-flow.md#optimize-data-flow).

> - In Camunda 8.8+, the Camunda Exporter and the Elasticsearch exporter run in the same thread within the broker, so Optimize export competes directly with core platform export for throughput.

Author: ```suggestion
- In Camunda 8.8+, the Camunda Exporter and the Elasticsearch exporter run in the same thread within the broker, so Optimize data-pipeline competes directly with core platform exporting for throughput.
```

- The overhead is **not proportional to throughput.** It scales with process model complexity (multi-instance and call activities) and variable volume. At a realistic workload where Optimize-enabled and Optimize-disabled clusters reached identical throughput with zero backpressure, the Optimize-enabled cluster still consumed **around 3.4x more Elasticsearch CPU.** Budget for this even at comfortable throughput.

#### What Optimize affects

At a realistic workload, Optimize enabled vs. disabled:

- **Elasticsearch CPU:** around 3.4x higher.

> - **Elasticsearch disk:** around 3.6x more total data. A multiplier scales better across cluster sizes and retention periods than an absolute figure.

Author: ```suggestion
- **Elasticsearch disk:** around 3.6x more total data.
```

- **Throughput:** 25-50% lower on the same hardware.
- **Write-to-exporting latency:** around 2.6x higher.
- **Backpressure at maximum load:** around 45% with Optimize vs. 35% without.

> - **Individual import latency:** increases approximately linearly with payload size (for example, an 11 KB realistic payload vs. a 0.5 KB typical one).

Author: ```suggestion
- **Individual import latency:** increases approximately linearly with payload size.
```

> - **Report loading times:** increase approximately linearly as historical data accumulates.

Author: ```suggestion
- **Report loading times:** increase approximately linearly with the complexity of data (like process instances and variables) and as historical data accumulates 
```

> Memory is not a meaningful differentiator.

Author: ```suggestion
Memory of the secondary storage is not a meaningful differentiator to improve performance.
```

:::note
**Watch Optimize import lag.** When Optimize's importer falls behind the export rate, two problems can appear:

- **Optimize's analytics indices grow.** Optimize keeps one document per process instance and can only apply retention-based cleanup once its importer has processed the instance's completion. While the importer lags, completions are recorded late, cleanup is deferred, and Optimize's own indices grow beyond their steady-state size.
- **Data can be missed.** The raw exporter indices are cleaned up on the Elasticsearch/OpenSearch retention schedule. If the importer falls far enough behind, those records are deleted before Optimize imports them, and that data never reaches Optimize. This Exporter-Importer hazard is exactly what the 8.8 Camunda Exporter architecture removed for Operate and Tasklist.

If you see persistent import lag, raise the import throughput (see mitigations below).
:::

#### Mitigations

##### Keep variables out of Optimize (highest impact, lowest risk)

> Variables dominate Optimize's storage and CPU cost: Optimize stores a variable roughly **14x more expensively than the raw export does** (around 29x for high-cardinality string variables), so almost the entire cost lives in Optimize's analytics indices. There are two levers:

Author: ```suggestion
Variables dominate Optimize's storage and CPU costs on secondary storage: Optimize stores a variable roughly **14x more expensively than the raw export** (around 29x for high-cardinality string variables), so almost the entire cost lives in Optimize's analytics indices. There are two levers:
```

Variables dominate Optimize's storage and CPU costs on secondary storage: Optimize stores a variable roughly **14x more expensively than the raw export** (around 29x for high-cardinality string variables), so almost the entire cost lives in Optimize's analytics indices. There are three levers, from most to least aggressive:

> - **Stop exporting variables entirely _(Camunda 8.9+)_.** Set `index.variable: false` at the Elasticsearch/OpenSearch exporter to drop all variable records. This is the only lever that also recovers throughput because the exporter write path is the bottleneck at maximum load.

Author: @afgambin I removed the markers because this was not fully true - as this was possible with 8.8 as well :)

> - **Stop exporting variables entirely.** Set `camunda.data.exporters.elasticsearch.args.index.variable: false` (OpenSearch: `camunda.data.exporters.opensearch.args.index.variable: false`) at the exporter to drop all variable records. This is the only lever that also recovers throughput because the exporter write path is the bottleneck at maximum load.

Author: In addition I added Opensearch config here
